src/Microsoft.CSharp/tests/CSharpBinaryOperationTests.cs

```diff
@@ -0,0 +1,535 @@
+            dynamic dY = y;
+            dX <<= dY;
+            Assert.Equal(x << y, dX);
+        }
+
+        [Theory, MemberData(nameof(CrossJoinInt32))]
+        public void ModuloInt32Assign(int x, int y)
+        {
+            dynamic dX = x;
+            dynamic dY = y;
+            if (y == 0)
+                Assert.Throws<DivideByZeroException>(() => dX %= dY);
+            else if (y == -1 && x == int.MinValue)
+                Assert.Throws<OverflowException>(() => dX %= dY);
+            else
+            {
+                dX %= dY;
+                Assert.Equal(x % y, dX);
+            }
+        }
+
+        [Theory, MemberData(nameof(CrossJoinInt32))]
+        public void MultiplyInt32Assign(int x, int y)
+        {
+            dynamic dX = x;
+            dynamic dY = y;
+            dX *= dY;
+            Assert.Equal(x * y, dX);
+        }
+
+        [Theory, MemberData(nameof(CrossJoinInt32))]
+        public void MultiplyOvfInt32Assign(int x, int y)
+        {
+            dynamic dX = x;
+            dynamic dY = y;
+            int result;
+            try
+            {
+                result = checked(x * y);
+            }
+            catch (OverflowException)
+            {
+                Assert.Throws<OverflowException>(() => checked(dX *= dY));
+                return;
+            }
+
+            dX *= dY;
+            Assert.Equal(result, dX);
+        }
+
+
+        [Theory, MemberData(nameof(CrossJoinInt32))]
+        public void OrInt32Assign(int x, int y)
+        {
+            dynamic dX = x;
+            dynamic dY = y;
+            dX |= dY;
+            Assert.Equal(x | y, dX);
+        }
+
+        [Theory, MemberData(nameof(CrossJoinInt32))]
+        public void RightShiftInt32Assign(int x, int y)
+        {
+            dynamic dX = x;
+            dynamic dY = y;
+            dX >>= dY;
+            Assert.Equal(x >> y, dX);
+        }
+
+        [Theory, MemberData(nameof(CrossJoinInt32))]
+        public void SubtractInt32Assign(int x, int y)
+        {
+            dynamic dX = x;
+            dynamic dY = y;
+            dX -= dY;
+            Assert.Equal(x - y, dX);
+        }
+
+        [Theory, MemberData(nameof(CrossJoinInt32))]
+        public void SubtractOvfInt32Assign(int x, int y)
+        {
+            dynamic dX = x;
+            dynamic dY = y;
+            int result;
+            try
+            {
+                result = checked(x - y);
+            }
+            catch (OverflowException)
+            {
+                Assert.Throws<OverflowException>(() => checked(dX -= dY));
+                return;
+            }
+
+            checked
+            {
+                dX -= dY;
+            }
+            Assert.Equal(result, dX);
+        }
+
+        [Theory, MemberData(nameof(CrossJoinDouble))]
+        public void AddDouble(double x, double y)
+        {
+            dynamic dX = x;
+            dynamic dY = y;
+            Assert.Equal(x + y, dX + dY);
```

Comment:
I noticed that, but wonder if it matters, especially since I'll be explicitly covering the same paths after tonight or tomorrow.
Aside from it being a bit less performant the only downside is it blowing up particularly spectacularly if we introduce a bug, so we'll just not do that ;) The same thing is true elsewhere; I had a really bizarre output once when I tried an optimisation too far in the Linq project.